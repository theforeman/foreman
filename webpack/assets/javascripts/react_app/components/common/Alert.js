import React from 'react';
import Icon from './Icon';
import { ALERT_CSS } from '../../constants';

const Alert = (props) => {
  let classNames = ALERT_CSS[props.type];

  if (props.css) {
    classNames += ' ' + props.css;
  }

  if (props.dismissable) {
    classNames += ' ' + ALERT_CSS.dismissable;
  }

  const closeBtn = (
    <button className="close" data-dismiss="alert" aria-hidden="true">
      <Icon type="close" />
    </button>
  );

  return (
    <div className={classNames}>
      {props.dismissable && closeBtn}
      {props.children}
    </div>
  );
};

Alert.defaultProps = { dismissable: false, css: null, type: 'info' };
export default Alert;
