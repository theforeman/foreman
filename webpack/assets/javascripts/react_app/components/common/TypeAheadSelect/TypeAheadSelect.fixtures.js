import Immutable from 'seamless-immutable';

export const id = 'test_type_ahead_select';
export const options = ['option1', 'option2'];
export const selected = ['option2'];
export const initialState = Immutable({ typeAheadSelect: {} });
export const populatedState = Immutable({
  typeAheadSelect: {
    test_type_ahead_select: {
      options,
      selected,
    },
  },
});

export const props = {
  id,
  options,
  selected,
};
