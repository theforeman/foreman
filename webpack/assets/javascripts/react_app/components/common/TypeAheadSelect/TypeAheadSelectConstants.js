export const INIT = 'TYPEAHEAD_INIT';
export const UPDATE_OPTIONS = 'TYPEAHEAD_UPDATE_OPTIONS';
export const UPDATE_SELECTED = 'TYPEAHEAD_UPDATE_SELECTED';
