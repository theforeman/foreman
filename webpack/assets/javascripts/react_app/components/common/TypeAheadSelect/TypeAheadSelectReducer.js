import Immutable from 'seamless-immutable';
import {
  INIT,
  UPDATE_OPTIONS,
  UPDATE_SELECTED,
} from './TypeAheadSelectConstants';

const initialState = Immutable({});

export default (
  state = initialState,
  { type, payload: { id, options, selected } = {} }
) => {
  switch (type) {
    case INIT:
      return state.setIn([id], {
        ...state[id],
        options,
        selected,
      });
    case UPDATE_OPTIONS:
      return state.setIn([id], {
        ...state[id],
        options,
      });
    case UPDATE_SELECTED:
      return state.setIn([id], {
        ...state[id],
        selected,
      });
    default:
      return state;
  }
};
