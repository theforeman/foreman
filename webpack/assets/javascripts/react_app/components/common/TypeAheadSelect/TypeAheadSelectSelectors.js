import Immutable from 'seamless-immutable';

export const mapSelected = selected => selected.map(item => item.label || item);

const selectTypeAheadSelect = ({ typeAheadSelect }, id) =>
  typeAheadSelect[id] || {};

export const selectTypeAheadSelectExists = ({ typeAheadSelect }, id) =>
  !!typeAheadSelect[id];

export const selectOptions = (state, id) => {
  const typeAhead = selectTypeAheadSelect(state, id);
  const options = typeAhead.options || [];
  return Immutable.isImmutable(options) ? options.asMutable() : options;
};

export const selectSelected = (state, id) =>
  selectTypeAheadSelect(state, id).selected;
