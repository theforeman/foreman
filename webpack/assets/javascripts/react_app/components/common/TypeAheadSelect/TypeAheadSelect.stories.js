import React from 'react';
import TypeAheadSelect from '.';
import Story from '../../../../../..//stories/components/Story';
import storeDecorator from '../../../../../..//stories/storeDecorator';

export default {
  title: 'Components/TypeAheadSelect',
  decorators: [storeDecorator],
  component: TypeAheadSelect,
};

export const defaultTypeAhead = () => (
  <Story>
    <TypeAheadSelect
      id="default_typeahead"
      options={['Bear', 'Walrus']}
      selected={['Walrus']}
    />
  </Story>
);

export const withPlaceholderText = () => (
  <Story>
    <TypeAheadSelect
      id="placeholder_typeahead"
      options={['Bear', 'Walrus']}
      placeholder="Inspirational placeholder"
    />
  </Story>
);

export const multipleSelections = () => (
  <Story>
    <TypeAheadSelect
      id="multiple_typeahead"
      options={['Bear', 'Walrus']}
      multiple
    />
  </Story>
);

export const freehandSelections = () => (
  <Story>
    <div>New options can be freely added that are not present in options</div>
    <TypeAheadSelect
      id="freehand_typeahead"
      options={['Bear', 'Walrus']}
      allowNew
    />
  </Story>
);

export const withClearButton = () => (
  <Story>
    <TypeAheadSelect
      id="clearbutton_typeahead"
      options={['Bear', 'Walrus']}
      selected={['Bear']}
      clearButton
    />
  </Story>
);
