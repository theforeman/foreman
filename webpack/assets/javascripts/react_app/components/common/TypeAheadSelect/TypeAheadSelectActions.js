import {
  INIT,
  UPDATE_OPTIONS,
  UPDATE_SELECTED,
} from './TypeAheadSelectConstants';
import { mapSelected } from './TypeAheadSelectSelectors';

export const initialUpdate = (options, selected, id) => ({
  type: INIT,
  payload: {
    id,
    options,
    selected,
  },
});

export const updateOptions = (options, id) => ({
  type: UPDATE_OPTIONS,
  payload: {
    id,
    options,
  },
});

export const updateSelected = (selected, id) => ({
  type: UPDATE_SELECTED,
  payload: {
    id,
    selected: mapSelected(selected),
  },
});
