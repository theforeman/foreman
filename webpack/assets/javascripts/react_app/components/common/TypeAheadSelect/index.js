import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { TypeAheadSelect } from 'patternfly-react';
import { initialUpdate, updateSelected } from './TypeAheadSelectActions';
import {
  selectTypeAheadSelectExists,
  selectOptions,
  selectSelected,
} from './TypeAheadSelectSelectors';
import reducer from './TypeAheadSelectReducer';

const ConnectedTypeAheadSelect = ({
  id,
  options,
  selected,
  allowNew,
  multiple,
  placeholder,
  defaultInputValue,
  clearButton,
  inputProps,
}) => {
  const dispatch = useDispatch();
  const exists = useSelector(state => selectTypeAheadSelectExists(state, id));

  useEffect(() => {
    if (!exists) {
      dispatch(initialUpdate(options, selected, id));
    }
  }, [dispatch, exists, options, selected, id]);

  const _selected = useSelector(state => selectSelected(state, id));
  const _options = useSelector(state => selectOptions(state, id));
  const onChange = items => dispatch(updateSelected(items, id));

  return (
    <TypeAheadSelect
      id={id}
      options={_options}
      selected={_selected}
      allowNew={allowNew}
      multiple={multiple}
      placeholder={placeholder}
      defaultInputValue={defaultInputValue}
      clearButton={clearButton}
      inputProps={inputProps}
      onChange={onChange}
    />
  );
};

ConnectedTypeAheadSelect.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  options: PropTypes.array,
  selected: PropTypes.array,
  allowNew: PropTypes.bool,
  multiple: PropTypes.bool,
  placeholder: PropTypes.string,
  defaultInputValue: PropTypes.string,
  clearButton: PropTypes.bool,
  inputProps: PropTypes.object,
};

ConnectedTypeAheadSelect.defaultProps = {
  options: [],
  selected: [],
  allowNew: false,
  multiple: false,
  placeholder: '',
  defaultInputValue: '',
  clearButton: false,
  inputProps: {},
};

export default ConnectedTypeAheadSelect;

export const reducers = { typeAheadSelect: reducer };
