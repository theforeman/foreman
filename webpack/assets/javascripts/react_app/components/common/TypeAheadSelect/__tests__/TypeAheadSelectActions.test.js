import { testActionSnapshotWithFixtures } from '../../../../common/testHelpers';
import {
  initialUpdate,
  updateOptions,
  updateSelected,
} from '../TypeAheadSelectActions';
import { id, options, selected } from '../TypeAheadSelect.fixtures';

const fixtures = {
  'initializes defaults': () => initialUpdate(options, selected, id),
  'updates options': () => updateOptions(options, id),
  'updates selections': () => updateSelected(selected, id),
};

describe('TypeAheadSelectActions', () =>
  testActionSnapshotWithFixtures(fixtures));
