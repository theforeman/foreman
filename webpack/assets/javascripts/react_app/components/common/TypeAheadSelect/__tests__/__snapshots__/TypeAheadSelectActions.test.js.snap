// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeAheadSelectActions initializes defaults 1`] = `
Object {
  "payload": Object {
    "id": "test_type_ahead_select",
    "options": Array [
      "option1",
      "option2",
    ],
    "selected": Array [
      "option2",
    ],
  },
  "type": "TYPEAHEAD_INIT",
}
`;

exports[`TypeAheadSelectActions updates options 1`] = `
Object {
  "payload": Object {
    "id": "test_type_ahead_select",
    "options": Array [
      "option1",
      "option2",
    ],
  },
  "type": "TYPEAHEAD_UPDATE_OPTIONS",
}
`;

exports[`TypeAheadSelectActions updates selections 1`] = `
Object {
  "payload": Object {
    "id": "test_type_ahead_select",
    "selected": Array [
      "option2",
    ],
  },
  "type": "TYPEAHEAD_UPDATE_SELECTED",
}
`;
